cmd/admin-handlers.go

```diff
@@ -415,6 +416,136 @@ func (a adminAPIHandlers) PerfInfoHandler(w http.ResponseWriter, r *http.Request
 	return
 }
 
+// LockEntry holds information about client requesting the lock,
+// servers holding the lock, source on the client machine,
+// ID, type(read or write) and time stamp.
+type LockEntry struct {
+	Timestamp  time.Time `json:"time"`       // Timestamp set at the time of initialization.
+	Resource   string    `json:"resource"`   // Resource contains info like bucket, object etc
+	Type       string    `json:"type"`       // Bool whether write or read lock.
+	Source     string    `json:"source"`     // Source which created the lock
+	ServerList []string  `json:"serverlist"` // RPC path of servers issuing the lock.
+	Owner      string    `json:"owner"`      // RPC path of client claiming lock.
+	ID         string    `json:"id"`         // UID to uniquely identify request of client.
+}
+
+func newLockEntry(l lockRequesterInfo, resource, server string) *LockEntry {
+	entry := &LockEntry{Timestamp: l.Timestamp, Resource: resource, ServerList: []string{server}, Owner: l.Node, Source: l.Source, ID: l.UID}
+	if l.Writer {
+		entry.Type = "Write"
+	} else {
+		entry.Type = "Read"
+	}
+	return entry
+}
+
+// LockEntries - To sort the locks
+type LockEntries []LockEntry
+
+func (l LockEntries) Len() int {
+	return len(l)
+}
+
+func (l LockEntries) Less(i, j int) bool {
+	return l[i].Timestamp.Before(l[j].Timestamp)
+}
+
+func (l LockEntries) Swap(i, j int) {
+	l[i], l[j] = l[j], l[i]
+}
+
+func topLockEntries(lockMaps []LocksResponse) LockEntries {
+	const listCount int = 10
+	entryMap := make(map[string]*LockEntry)
+	for _, lockMap := range lockMaps {
+		for k, v := range lockMap.LockResponse {
+			for _, lockReqInfo := range v {
+				if val, ok := entryMap[lockReqInfo.UID]; ok {
+					val.ServerList = append(val.ServerList, lockMap.Addr)
+				} else {
+					entryMap[lockReqInfo.UID] = newLockEntry(lockReqInfo, k, lockMap.Addr)
+				}
+			}
+		}
+	}
+	var lockEntries LockEntries
+	for _, v := range entryMap {
+		lockEntries = append(lockEntries, *v)
+	}
+	sort.Sort(sort.Reverse(lockEntries))
+	if len(lockEntries) > listCount {
+		lockEntries = lockEntries[:listCount]
+	}
+	return lockEntries
+}
+
+// LocksResponse holds server information result of one node
+type LocksResponse struct {
+	Error        string       `json:"error"`
+	Addr         string       `json:"addr"`
+	LockResponse TopLocksResp `json:"locksresponse"`
+}
+
+func lockMapCopy(lockMap map[string][]lockRequesterInfo) map[string][]lockRequesterInfo {
+	lockCopy := map[string][]lockRequesterInfo{}
+	for k, v := range lockMap {
+		var lockSlice []lockRequesterInfo
+		for _, lockInfo := range v {
+			lockSlice = append(lockSlice, lockInfo)
+		}
+		lockCopy[k] = lockSlice
+	}
+	return lockCopy
+}
```

Comment:
globalLockServer.ll.lockMap should not be touched without holding locks, it can be done like this:
```
func (l *localLocker) DupLockMap() map[string][]lockRequesterInfo {
        l.mutex.Lock()
        defer l.mutex.Unlock()

        lockCopy := make(map[string][]lockRequesterInfo)
        for k, v := range lockMap {
                var lockSlice []lockRequesterInfo
                for _, lockInfo := range v {
                        lockSlice = append(lockSlice, lockInfo)
                }
                lockCopy[k] = lockSlice
        }
        return lockCopy
}
```

Reply:
done